torch/_inductor/scheduler.py

```diff
@@ -4207,6 +4214,80 @@ def get_graph_partition_signature(
 
         return signatures[::-1]
 
+    def reorder_for_partition_based_on_devices(
+        self, nodes: list[BaseSchedulerNode]
+    ) -> list[BaseSchedulerNode]:
+        """
+        Given topologically sorted `nodes`, reorder by devices while
+        respecting dependencies. Specifically, when a node reads from
+        another device, execute all leading nodes on that device since
+        we already have a graph partition for that device.
+        """
+
+        result: list[BaseSchedulerNode] = []
+        device_to_nodes: dict[Optional[torch.device], list[BaseSchedulerNode]] = (
+            collections.defaultdict(list)
+        )
+
+        def visit(node: BaseSchedulerNode) -> None:
+            node_device = node.get_device()
+            for dep in node.unmet_dependencies:
+                if (
+                    buffer := self.name_to_buf.get(dep.name, None)
+                ) and buffer.get_device() != node_device:
+                    input_device = buffer.get_device()
+                    # the dependent node may have been processed
+                    if input_device_nodes := device_to_nodes.pop(input_device, None):
+                        result.extend(input_device_nodes)
+            device_to_nodes[node_device].append(node)
+
+        for node in nodes:
+            visit(node)
```

Comment:
I dont follow this.. we visit each node, and in each node we add `device_to_nodes[node_device].append(node)`. what is result supposed to represent ?

we always add each node in line 4242. and then there are some nodes we add twice. then we add those nodes to result.